aten/src/ATen/native/nested/NestedTensorMath.cpp

```diff
@@ -576,55 +573,27 @@ Tensor& NestedTensor_mul__Tensor(Tensor& self, const Tensor& other) {
 }
 
 Tensor select_nested(const Tensor& self, int64_t dim, int64_t index) {
+  auto self_ptr = get_nested_tensor_impl(self);
+  int64_t positive_dim = at::maybe_wrap_dim(dim, self_ptr->dim());
   TORCH_CHECK(
-    dim == 0,
+    positive_dim == 0,
     "NestedTensor can only be selected along dimension 0 ",
     "got dimension ", dim, " instead."
   );
-  auto self_ptr = get_nested_tensor_impl(self);
-  // buffer contains the underlying data in a contiguous vector
-  const at::Tensor & buffer = self_ptr->get_buffer();
-  int64_t numel = buffer.numel();
-  TORCH_CHECK(
-    numel > 0,
-    "cannot index an empty nested tensor."
-  );
-  // nested_tensor[i] = i-th original tensor
-  int64_t ntensors = *(self_ptr->opt_size(0));
-  int64_t positive_index = at::maybe_wrap_dim(index, ntensors);
-  // determine the memory segment of the i-th original tensor
-  Tensor sizemat = get_nested_size_tensor(self);
-  int64_t original_dim = sizemat.size(1);
-  const int64_t * sizemat_ptr = sizemat.data_ptr<int64_t>();
-  // start of the segment
-  int64_t start = 0, sizemat_offset = 0;
-  for (int64_t i = 0; i < positive_index; i++) {
-    int64_t row_product = sizemat_ptr[sizemat_offset];
-    sizemat_offset++;
-    for (int64_t j = 1; j < original_dim; j++) {
-      row_product *= sizemat_ptr[sizemat_offset];
-      sizemat_offset++;
-    }
-    start += row_product;
-  }
-  // btw determine the shape of the i-th original tensor
-  IntArrayRef shape(sizemat_ptr + sizemat_offset, sizemat_ptr + sizemat_offset + original_dim);
-  // stop of the segment
-  int64_t stop;
-  if (positive_index == ntensors - 1) {
-    stop = numel;
+  int64_t ntensors = 0;
+  if (self_ptr->get_nested_size_tensor().dim() == 2) {
+    ntensors = *(self_ptr->opt_size(0));
   }
-  else {
-    int64_t row_product = sizemat_ptr[sizemat_offset];
-    sizemat_offset++;
-    for (int64_t j = 1; j < original_dim; j++) {
-      row_product *= sizemat_ptr[sizemat_offset];
-      sizemat_offset++;
-    }
-    stop = start + row_product;
-  }
-  // extract the memory segment then reshape to the original shape
-  return buffer.slice(0, start, stop).view(shape);
+  TORCH_CHECK_INDEX(
+      index >= -ntensors && index < ntensors,
+      "index ", index,
+      " is out of bounds for dimension 0 with size ", ntensors);
+  int64_t positive_index = index < 0 ? index + ntensors : index;
```

Comment:
nit: could use `maybe_wrap_dim()` here as well

Reply:
The reason why I changed from `maybe_wrap_dim()` to this `TORCH_CHECK_INDEX` (mimiced from tensor select) is:
* if we have an empty nested tensor (`ntensors = 0`), then `index = 0` or `index = -1` will pass the bound check then cause segmentation fault
* to be more specific, say we have `ntensors = 0` then pass `index = 10`, the error message we get is `IndexError: Dimension out of range (expected to be in range of [-1, 0], but got 10)`

Reply:
oh I see, it doesn't allow for 0 dims. Thanks for the context!

Reply:
edge case is always painful :anguished: